internals-js/src/federation.ts

```diff
@@ -486,6 +497,18 @@ export class FederationMetadata {
     return this.sharingPredicate()(field);
   }
 
+  isKeyField(field: FieldDefinition<any>): boolean {
+    return this.keysPredicate()(field);
+  }
+
+  addSyntheticExternal(coordinate: string) {
```

Comment:
I haven't add time to check all the recent changes yet, but wanted to give early feedback on this one.

I'm worried about this, because this change what the `FederationMetadata` abstraction is, and in a way that I worry will hurt maintenance. What I mean here is that so far, `FederationMetadata` is kind of "stateless" in the sense that it's only meant to provide convenient access to federation-related information, and it does pre-compute/cache things, but it can all be recomputed from the schema and `FederationMetadata` has no information of its own (and btw, I reckon that the name, `FederationMetadata`, may not be ideal in that it does not convey that very well).

Importantly, that means that if we clone or print a schema, we don't care about `FederationMetadata`, there is no real information in it.

This change that, and that worries me.

Fwiw, I see 2 potential alternatives:
1. we modify merging to only include an overridden field as external in the subgraph if it is used by a key/provides/requires/interface. Essentially, the only place when `isSyntheticExternal` is used is to bypass the "is that external used?" check, so we can simply avoid adding the overridden field at all if it's not used. It's the solution that I initially suggested because it's imo the simplest one.
2. if we instead want to always keep overridden fields in the supergraph and have them extracted as special "synthetic external", then I'd prefer creating a new specific `@federation__internal__syntheticExternal` directive to mark the field so that all information remain contained in the schema itself. I wouldn't vote for this option as I think the previous one is simpler, at least for now, and good enough, but I wouldn't oppose it either.

Reply:
Ok, went with option 1. Let me know what you think.